library/core/src/num/mod.rs

```diff
@@ -1074,13 +1089,15 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
         // `u8::MAX` is `ff` - any str of len 2 is guaranteed to not overflow.
         // `i8::MAX` is `7f` - only a str of len 1 is guaranteed to not overflow.
         macro_rules! run_unchecked_loop {
-            ($unchecked_additive_op:expr) => {
-                for &c in digits {
+            ($unchecked_additive_op:expr) => {{
+                let mut i = 0; // C-style iteration due to lack of const_for.
+                while let Some(&c) = digits.get(i) {
```

Comment:
Is this adds\removes bound checks, and generally, perf changes?

This ( and lower one) can be rewritten without `i` like `while let Some((first, last)) = digits.split_first()`

Reply:
Looks like one bounds check per iteration to me (in `get()` only) and [Godbolt suggests](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahSAVyVFyKxqiIEh1ZpgDC6embZMpATnLOAGQImbAA5TwAjbFIQADZyAAd0JWIHJjcPL19E5NShIJDwtiiY%2BJtsOzSRIhZSIgzPbx4/csqhatqiArDI6LjrGrqGrOaBzu6ikriASmt0M1JUTi5sdSJopgBqAFIAdgAhba0AQU2zzZotwXQITU2zAA5p7YBmQ5O9gBEj45%2BEswiFy26ggERAOwATLFtgBWfaPWGfaY7A4/c6bRhETZsMxYgg7F6fTZaV7vU7nADuSAY2Ax2CxInQHAg2yhEQAnutka8iREAHTAekQAjc1EndHosxMJQsGi0vb7C7oG4crko75vNESs741mHQmbHikrVnL4/L5cWb0bgw/jeLg6cjobgAJQsWKU80W8ohLz45CI2ktswA1iAXhC%2BRCHloeA9YjwXrtYtIXg8fDDDNxpPw2OHdnzpLs49JYg9S%2BWy5n7Y7nVx%2BEoQFoA0HZnBYCgMDh8MQyJRqHRGKwONx/YJhGIJJwZHJhMo1JoHXoQBDDMZTO7rNhbPZHBBnMNvKvAsEesU%2Bn4kild%2Bl3I0Vzlr2kJr0Yi1txUbx0hneRlud1UgxdKekx9DwYw/pkTQQcBhSvlIsyegsSzcKs6ykFsComkCSo3HcjzPJqHy7N8HwnP8gKXJsIJgpC0JwgiMJIiiZLooIpB0aqtLBJsgJYeK2r3NKsrygcuGgpy2CihqrHnGaxGfJaWZcLa5A1vwdZupYmxId6kIRvwgZLtMVrZrm4ZaHyMKxDCLwwjC8YPC8zm7FosjqU63ANk2LbGeQ7bIGgWB4IQJAUFQtA0sOyxjvIk6SDO46KCoGhBuQ%2BgvGuJggOYlj/p%2BaROEwri/ke/jFS%2B54xDwzZXnkt5QQYdU3pVUyZq0X5AYeD4dYB4wgfBNUwd14GWP1cFVSAMKIV6KErGsGwsdhVHXLc4IEcaCnmuRAI4SCOJYrRrL0fCDyIqKslnFSNJ0lisL7Fx5CbKQ2DaQAApsfJ8oiBK8ktAnalKMpyix4kAFRcdJm3ktqfEGi9ljQ%2Bi8m/CRSnWipdppZp7o6bNPrOYZrazEg2AsDgMQQKZXA5uQeZ%2BnyPixBCuw%2BLs0bSA8zTSNI1bY151g%2BUZOgmeQYbSD4Vl%2Bj4Ly86zsQyw5ykvFjS6efWvki/5iABRAnZMgkNLhdQGBsIbjAxKQPDM82g7oY2oJpREwS1Oyo78KbHDCAA8kw9Bu2rOBsOukiBwQL2VAAbq9aWrBUuIxfwwTrBjjr0AQESkK7bg4O7AakAQea8EpNBGMASgAGoENgFLewkzB50l8XTrISXzqlav6OBRjZblVjpxEjaQLM6AJDejZcAAtC4myj0Qk%2BMNH9A8ir%2BVtN4%2B7FSN5WYK1YGPvV2/Nc%2BA2TaNH7r9%2B9SlQYvXtEBe/VcNN%2BjQ/p9TDwM3IdO%2BfYIn1OqQ8nWPuqBDR8mZnyLQmwIBaSsJsHsYV9LgU2G4M2RskHTCJsZEmZMKbUFDFNSyrkIRaB8FzF4bkYRM0yhjWm9MCx8zVnWbyzZhZKV1vrNBFt%2BwQFNubPoVsbYCAYPbagYI1bO1YKQAO/pPbMCIL7f2aUg4h2WI6QgEd7DRwno6OOqAE552TtuNKA8s7SJzmowyBci58FmKXFg5cq41zrg3Yus5RDiASq3eQ7dFyOn0KuHuG48oDyHlTJ0Y80gT2nrPMeC9sBLxXg2C%2BN4iolUauBE8E0P4HxvEfXILV377zvkwK%2B28SlX0frfLqL8YJVM/nMb%2BD4iAvX/spQB/MuAgLARAqBMDNzwNCmQfSq4UEG3QayCMmDNbBnIKTcmfRwm0PMhQvkMgfBaBhOszZ2zdhqU6SwmZbYdacP4cbXh4yLY5WtiQ4R9BRGOwkS7aRec5E%2Bz9gHdR2Bg4mFDuo8OBVtGx3UPHdYhjhDGLVqY7OwU0otMLu7OxZdK7V1rvXe0sUJyeJbu43xaUMpZWCf3DOYSR6RKEBPZJAE9wHlqVks8OTj5CHyU%2BfIRSn4VJqRktenVxoMuKVy%2B8r8%2BWgWql/b04Zf5tIxh0ph3Buk8HAZGPpsCnoIOGZMzKYyuHRH0i8aZbDRbzLwUssydNwyRjsqQzmMJrUPC2fsuVGtGysOJmLQhhYyxaBstIEhFCUwQlkBjVeQCBaGuphCVWtYw1uujqQFIjhpBAA) that the version with `i` compiles to the same as `for &c in digits`, but the pattern matching one doesn't get optimized so well.

Reply:
In your link you missed `-C opt-level=3` for 3rd code example.

Reply:
Thanks - they are all the same then. Personally I think the version with `i` is less complicated to read but no firm preference.